src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java

```diff
@@ -278,21 +278,36 @@ private void processIdent(DetailAST ast) {
         final DetailAST parent = ast.getParent();
         final int parentType = parent.getType();
 
-        final boolean isPossibleDotClassOrInMethod = parentType == TokenTypes.DOT
-                || parentType == TokenTypes.METHOD_DEF;
-
-        final boolean isQualifiedIdent = parentType == TokenTypes.DOT
-                && !TokenUtil.isOfType(ast.getPreviousSibling(), TokenTypes.DOT)
-                && ast.getNextSibling() != null;
-
         if (TokenUtil.isTypeDeclaration(parentType)) {
             currentFrame.addDeclaredType(ast.getText());
         }
-        else if (!isPossibleDotClassOrInMethod || isQualifiedIdent) {
-            currentFrame.addReferencedType(ast.getText());
+        else {
+            final boolean isPossibleDotClassOrInMethod = parentType == TokenTypes.DOT
+                    || parentType == TokenTypes.METHOD_DEF || parentType == TokenTypes.METHOD_REF;
+            final boolean isQualifiedIdent = parentType == TokenTypes.DOT
+                    && !TokenUtil.isOfType(ast.getPreviousSibling(), TokenTypes.DOT)
+                    && ast.getNextSibling() != null;
+            final boolean isQualifiedIdentFromMethodRef = isQualifiedIdentFromMethodRef(ast);
```

Comment:
It would be better if this method did not care qualified identifier is from method ref specifically. Can we just create a helper method like `isQualifiedIdentifier` and handle all qualified identifiers there, i.e. move logic from `isQualifiedIdent` and `isQualifiedIdentFromMethodRef` to this helper method?

Reply:
It would be good to make `processIdent` look like (psuedocode):
```java

// logic for booleans here

if (isTypeDeclaration) {
..
}
else if (!isInClassOrMethod || isQualifiedIdentifier(ast)) {
...
}
```